kafka/src/main/java/com/metamx/tranquility/kafka/writer/SynchronizedEventWriter.java

```
@@ -0,0 +1,121 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.metamx.tranquility.kafka.writer;
+
+import com.metamx.common.logger.Logger;
+
+/**
+ * An EventWriter that tracks the number of events sent to it and provides a waitForFlush() method that blocks until
+ * all of the events have flushed (the callback for each event has been called). This is so that we can guarantee all
+ * events have been pushed to Druid before committing the Kafka offsets.
+ */
+public class SynchronizedEventWriter implements EventWriter
+{
+  private static final Logger log = new Logger(SynchronizedEventWriter.class);
+
+  private final EventWriter writer;
+
+  // used to wait for completion callback from all events after flushing before committing the offset to ZK
+  private final Object flushAwaitLock = new Object();
+  private int pendingEvents = 0; // Use under flushAwaitLock
+
+  public SynchronizedEventWriter(EventWriter writer)
+  {
+    this.writer = writer;
+    writer.init(
+        new EventWriter.EventWriterListener()
+        {
+          @Override
+          public void onSuccess(Object o)
+          {
+            decrementPending();
+          }
+
+          @Override
+          public void onFailure(Throwable throwable, Object o)
+          {
+            decrementPending();
+            log.error(throwable, "Failed to send event: %s", (o == null ? "null" : o.toString()));
+          }
+        }
+    );
+  }
+
+  private void incrementPending()
+  {
+    synchronized (flushAwaitLock) {
+      pendingEvents++;
+    }
+  }
+
+  private void decrementPending()
+  {
+    synchronized (flushAwaitLock) {
+      pendingEvents--;
+      if (pendingEvents == 0) {
+        flushAwaitLock.notifyAll();
+      }
+    }
+  }
+
+  @Override
+  public void init(EventWriterListener listener)
+  {
+    // do nothing
```

This should throw an exception as it appears intended to never actually be called. (the listener would be ignored)